### PR TITLE
[feat] Qwen 3 VL MOE with EP support

### PR DESCRIPTION
**Pull request by ngquangtrung57**
## Commit Message Convention

Please follow our standardized commit message format:

- `[feat]` - New features or functionality
- `[fix]` - Bug fixes
- `[docs]` - Documentation changes only
- `[style]` - Code style changes (formatting, missing semicolons, etc.)
- `[refactor]` - Code refactoring without changing functionality
- `[perf]` - Performance improvements
- `[test]` - Adding or updating tests
- `[chore]` - Maintenance tasks, dependency updates, etc.
- `[ci]` - CI/CD configuration changes

**Examples:**
- `[feat] add qwen omni iterable dataset support`
- `[fix] resolve bagel model configuration error`
- `[docs] update training guide with YAML examples`

See [CONTRIBUTING.md](../CONTRIBUTING.md) for more details.

## CI/CD Checks

Your PR will automatically run the following checks:

- **Linting**: Code formatting with `black` (line-length=120) and import sorting with `isort`
- Run `pre-commit run --all-files` locally to verify before pushing

## Checklist

- [ ] Follow commit message convention (see above)
- [ ] Run `pre-commit run --all-files` and ensure all checks pass
- [ ] Format your code with `black` (line-length=120) and `isort`
- [ ] Add unit tests for new functionality
- [ ] Update documentation as needed, including docstrings or example tutorials
- [ ] Ensure all CI/CD checks pass

### CODE REVIEW

File: src/lmms_engine/models/qwen3_vl_moe/qwen3_vl_moe_experts.py
Comment (kcz358) [inline src/lmms_engine/models/qwen3_vl_moe/qwen3_vl_moe_experts.py]: Qwen3 VL text moe already in nn.Parameters. No need to wrap another layer for it

https://github.com/huggingface/transformers/blob/66d57110f089789ae285cc9d54d3bf051123246b/src/transformers/models/qwen3_vl_moe/modeling_qwen3_vl_moe.py#L68-L127

File: src/lmms_engine/parallel/qwen3_vl_moe/parallelize.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+            stacked_gate_up_proj = torch.stack(gate_up_proj_weights, dim=0)
+            new_experts.gate_up_proj = nn.Parameter(stacked_gate_up_proj)
+
+            # Stack down_proj weights
+            down_proj_weights = [expert.down_proj.weight for expert in decoder_layer.mlp.experts]
+            stacked_down_proj = torch.stack(down_proj_weights, dim=0)
+            new_experts.down_proj = nn.Parameter(stacked_down_proj)
+
+            del decoder_layer.mlp.experts
+            decoder_layer.mlp.add_module("experts", new_experts)
```
Comment (kcz358) [inline src/lmms_engine/parallel/qwen3_vl_moe/parallelize.py]: Referred to the comments above, I think there is no need for unpacking and packing the parameters the moe layer in qwen3 vl moe? As it is already packed.

**Review by kcz358 — COMMENTED** [review]
Hi, thank you for all the contribution. I notice that most of the code are copying from qwen3 moe and qwen3 omni moe. After checking the modeling file in https://github.com/huggingface/transformers/blob/main/src/transformers/models/qwen3_vl_moe/modeling_qwen3_vl_moe.py#L68-L127, I wonder if we can reuse this so that make the code cleaning. 

For example, we don't need to unstack and stack the parameters everytime to save time. Additionally (I am not sure if it works), by after looking at the code, I notice that it use the similar api we want to achieve. It is also using routing weights and logits to calculate the hidden states, wondering if we can directly use ParallelStyle on it so that we only need to write the all2all in input_fn and output_fn and no need for monkey patch it

**Comment by ngquangtrung57** [comment]
If im not wrong, the current approach with EP we are using route the token outside of the current HF forward. And as it is also routing the token inside forward, then isn't it gonna be redundant

**Comment by kcz358** [comment]
The reason that we need to monkey patch the forward is because that we need to calculate each token's logits all at the same time. In 4.57.1 transformers version, they implement it in a for loop and expert mask way

https://github.com/huggingface/transformers/blob/8cb5963cc22174954e7dca2c0a3320b7dc2f4edc/src/transformers/models/qwen3_moe/modeling_qwen3_moe.py#L226-L264

which is hard to perform all to all. Thus we will need to patch our own logic and re-stack the modulelist parameters so that we can do grouped-mm-forward or for-loop-forward (our implementation to save memory). In qwen3 vl, you can see that the router weights, router indices, and hidden states are all being send to the experts

https://github.com/huggingface/transformers/blob/66d57110f089789ae285cc9d54d3bf051123246b/src/transformers/models/qwen3_vl_moe/modeling_qwen3_vl_moe.py#L130-L153

which if I understanding correctly, we can direct apply all2all in _input_fn and _output_fn and we no need to patch to logic for calculating the moe router or logits or expert forwarding logic. Possibly just to patch the forwarding logic in

https://github.com/huggingface/transformers/blob/66d57110f089789ae285cc9d54d3bf051123246b/src/transformers/models/qwen3_vl_moe/modeling_qwen3_vl_moe.py#L79-L127

so that we handle to local etc.

Interestingly, in current latest branch, I saw that they are changing the expert implementation for Qwen3 Moe to nn.Parameters implementation, which means we might also need to change our previous implementation also ......

**Comment by ngquangtrung57** [comment]
Sure, they also made same changes to Qwen 3 Omni MOE as well. Let's work together to refine the approach

**Review by kcz358 — APPROVED** [review]
